src/GafferArnoldUI/ArnoldLightVisualiser.cpp

```diff
@@ -0,0 +1,399 @@
+
+//////////////////////////////////////////////////////////////////////////
+// Network conversion helpers
+//////////////////////////////////////////////////////////////////////////
+
+// Re-connects any connections from oldName on oldShader to newName on newShader
+void remapOutputConnections( const InternedString &oldShader, const InternedString &oldName, const InternedString &newShader, const InternedString &newName, ShaderNetwork *network )
+{
+	ShaderNetwork::Parameter newSource( newShader, newName );
+
+	ShaderNetwork::ConnectionRange inputConnections = network->outputConnections( oldShader );
+	for( ShaderNetwork::ConnectionIterator it = inputConnections.begin(); it != inputConnections.end(); )
+	{
+		// Copy and increment now so we still have a valid iterator
+		// if we remove the connection.
+		const ShaderNetwork::Connection connection = *it++;
+
+		if( connection.source.name == oldName )
+		{
+			ShaderNetwork::Parameter dest( connection.destination.shader, connection.destination.name );
+			network->removeConnection( connection );
+			network->addConnection( ShaderNetwork::Connection( newSource, dest ) );
+		}
+	}
+
+	ShaderNetwork::Parameter out = network->getOutput();
+	if( out.shader == oldShader && out.name == oldName )
+	{
+		network->setOutput( newSource );
+	}
+}
+
+// Sets outName in outParams if inName is set in inParams.
+// Arnold data types not representable in OSL will be converted accordingly
+void aiParamToOslParam( const InternedString &inName,  const CompoundDataMap &inParams, const InternedString &outName, CompoundDataMap &outParams )
+{
+	const auto &it = inParams.find( inName );
+	if( it != inParams.end() )
+	{
+		if( const BoolData *boolData = dynamic_cast<const BoolData *>( it->second.get() ) )
+		{
+			outParams[ outName ] = new IntData( 1 ? boolData->readable() : 0 );
+		}
+		else
+		{
+			outParams[ outName ] = it->second;
+		}
+	};
+}
+
+// Attempts to create an OSL stand-in shader for the supplied Arnold shader.
+// Any parameters that exist on the stand-in shader will be populated by the
+// equivalent parameter on the source shader.  If a matching OSL shader doesn't
+// exist, then a nullptr is returned.  Because Arnold shaders have a single
+// output, but OSL shaders don't, outputName will be populated with the name of
+// the first output of the OSL stand-in shader
+ShaderPtr oslStandin( const Shader *arnoldShader, std::string &outputName )
+{
+	const std::string oslShaderName = "__viewer/__arnold_" + arnoldShader->getName();
+
+	const OSLQueryPtr query = g_oslQueryCache.get( OSLQueryCacheGetterKey( oslShaderName ) );
+	if( !query )
+	{
+		return nullptr;
+	}
+
+	ShaderPtr oslShader = new IECoreScene::Shader();
+	oslShader->setType( "osl:shader" );
+	oslShader->setName( oslShaderName );
+
+	bool foundOutput = false;
+
+	const IECore::CompoundDataMap &inParams = arnoldShader->parameters();
+	IECore::CompoundDataMap &outParams = oslShader->parameters();
+	for( size_t i = 0; i < query->nparams(); ++i )
```

Comment:
If we don't do this, we need a list in this code of all the possible collisions with OSL reserved words. As with the output topic, I preferred this code being dumb but solid, in the sense that the simple rule of "suffix collisions with underscore" is all that it needs to know. If there is a new collision in a future shader, then this will just 'work', rather than having to worry about maintaining the list.